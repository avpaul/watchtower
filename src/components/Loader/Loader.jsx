import React from 'react';

import './Loader.css';

/**
 * Loader UI Component
 *
 * @returns {JSX} React component
 */
const Loader = () => (
  <div className="loader__center">
    <div className="loader" />
  </div>
);

export default Loader;
