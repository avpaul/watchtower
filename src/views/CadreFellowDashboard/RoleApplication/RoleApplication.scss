@import '../../../assets/sass/abstract/variables';

%equal-properties {
  height: 17px;
  width: 287px;
  font-family: $font-primary;
  font-size: 13px;
  letter-spacing: 0.12px;
  line-height: 17px;
}

.advocacy {
  width: 100%;
  &__label {
    @extend %equal-properties;
    margin-bottom: 1.5em;
    color: #7a8691;
  }

  &__subtitle {
    @extend %equal-properties;
    margin-bottom: 0.5em;
    color: #333e44;
  }
}

.modal-header h1 {
  color: #344c5a;
}

.success {
  &__image {
    height: 148px;
    width: 115px;
    color: #7a8691;
    font-family: 'Apple Color Emoji';
    font-size: 113px;
    letter-spacing: 1.03px;
    line-height: 148px;
  }

  &__title {
    color: #344c5a;
    width: 100%;
    font-family: $font-primary;
    font-size: 24px;
    font-weight: 500;
    letter-spacing: -0.69px;
    line-height: 31px;
  }

  &__sutitle {
    height: 17px;
    width: 100%;
    color: #7a8691;
    font-family: $font-primary;
    font-size: 13px;
    letter-spacing: 0.12px;
    line-height: 17px;
  }

  &__application {
    text-align: center;
    width: 100%;
    margin-top: -2em;
    margin-bottom: -2em;
  }
}

#applicationBtn {
  width: 40%;
}

#buttonDisabled {
  background-color: #cccccc;
  color: #666666;
}
