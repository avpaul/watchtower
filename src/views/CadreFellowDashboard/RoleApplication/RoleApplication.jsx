import React, { Component } from 'react';
import PropTypes from 'prop-types';
import GenericModal from '../../../components/GenericModal';
import successImage from '../../../static/like.png';
import Loader from '../../../components/Loader/Loader';
import FormInputs from '../../../components/FormInputs';
import './RoleApplication.scss';

class RoleApplication extends Component {
  constructor(props) {
    super(props);
    this.state = {
      inputs: {},
      success: false,
      errorMessage: '',
      currentLength: 0,
      isValidLength: false
    };
  }

  componentDidUpdate(prevProps) {
    const { applications } = this.props;
    if (
      (prevProps.applications.loading !== applications.loading &&
        prevProps.applications.data !== applications.data) ||
      prevProps.applications.error !== applications.error
    ) {
      this.applicationStatus(applications);
    }
  }

  applicationStatus = applications => {
    const { error } = applications;

    if (!error) {
      this.setState({ success: true });
    } else if (error && error.application_reason) {
      const { application_reason: reason } = error;
      this.setState({ errorMessage: reason[0] });
    } else if (error && error === 'You have an active role') {
      this.setState({
        errorMessage: error,
        isValidLength: false
      });
    }
  };

  handleSubmit = () => {
    const { applyForRole, roleId, projectId, cycleId } = this.props;
    const {
      inputs: { description }
    } = this.state;

    const currentValue = description.getValue();
    const data = {
      application_reason: currentValue,
      cycle_id: cycleId
    };

    if (!this.evaluateLength(currentValue.trim())) {
      return this.setState({
        errorMessage: 'Spaces, really?',
        isValidLength: false
      });
    }

    return applyForRole(Number(projectId), Number(roleId), data);
  };

  handleClose = () => {
    const { history, roleId } = this.props;
    this.setState({ success: false });
    history.replace(`${roleId}`);
  };

  renderButton = ({ label, buttonProps = {} }) => (
    <button type="button" className="btn" {...buttonProps}>
      {label}
    </button>
  );

  footer = (success, button) => (
    <div className="modal-footer">
      {success
        ? ''
        : this.renderButton({
            label: 'Cancel',
            buttonProps: {
              'data-dismiss': 'modal'
            }
          })}
      {button}
    </div>
  );

  evaluateLength = value => {
    const isValidLength = value.length >= 50 && value.length <= 500;
    this.setState({
      isValidLength,
      currentLength: value.length,
      errorMessage: ''
    });
    return isValidLength;
  };

  sendApplicationButton = () => {
    const { success, isValidLength } = this.state;
    const { applications } = this.props;
    let button = this.renderButton({
      label: 'Send Application',
      buttonProps: {
        onClick: this.handleSubmit,
        id: 'applicationBtn',
        disabled: !isValidLength
      }
    });
    if (success) {
      button = this.renderButton({
        label: 'Close',
        buttonProps: {
          onClick: this.handleClose,
          'data-dismiss': 'modal'
        }
      });
    }

    return applications.loading ? (
      <Loader size="small" />
    ) : (
      this.footer(success, button)
    );
  };

  converToUnicode = hexValue => String.fromCodePoint(hexValue);

  renderBody = () => {
    const { roleInfo } = this.props;
    const { success, errorMessage, currentLength } = this.state;
    const applicationUnits = {
      subtitle: `You are applying as a ${roleInfo.name}`,
      selfAdvocateTitle: `Sell yourself [500 characters max]`,
      successTitle: 'Great! Thank you for applying',
      successSutitle: 'You should hear from us shortly'
    };

    let progressClass;
    let feedbackText;

    if (currentLength >= 0 && currentLength < 50) {
      progressClass = 'danger';
      feedbackText = `C'mon, don't sell yourself short!. ${this.converToUnicode(
        128527
      )}`;
    } else if (currentLength > 500) {
      progressClass = 'danger';
      feedbackText = `OK, maybe not your full CV, 10x Engineer. ${this.converToUnicode(
        128540
      )}`;
    } else {
      progressClass = 'success';
      feedbackText = `Yeah! Keep it nice, short and simple. ${this.converToUnicode(
        128076
      )}`;
    }

    return (
      <React.Fragment>
        {success ? (
          <div className="success__application">
            <p className="success__title">{applicationUnits.successTitle}</p>
            <img className="success__image" src={successImage} alt="..." />
            <p className="success__sutitle">
              {applicationUnits.successSutitle}
            </p>
          </div>
        ) : (
          <>
            <div className="advocacy">
              <p className="advocacy__label">{applicationUnits.subtitle}</p>
              <p className="advocacy__subtitle">
                {applicationUnits.selfAdvocateTitle}
              </p>
              <FormInputs.TextInput
                parent={this}
                name="description"
                type="textarea"
                testInput={this.evaluateLength}
              />
              {errorMessage && (
                <span className="error-message text-danger">
                  {errorMessage}
                </span>
              )}
            </div>
            <div className="d-flex w-100 word-count justify-content-between">
              <p className="feedback-text">{feedbackText}</p>
              <div
                className={`rounded-circle count-wrapper d-flex justify-content-center align-items-center ml-3 border-${progressClass} ml-3`}
              >
                <p className="count">{currentLength}</p>
              </div>
            </div>
          </>
        )}
      </React.Fragment>
    );
  };

  render() {
    const { success } = this.state;
    const { projectTitle } = this.props;
    const title = `Apply to ${projectTitle}`;

    return (
      <GenericModal
        id="applyForRoleModal"
        title={success ? '' : title}
        footer={this.sendApplicationButton()}
      >
        {this.renderBody()}
      </GenericModal>
    );
  }
}

RoleApplication.propTypes = {
  applications: PropTypes.shape().isRequired,
  applyForRole: PropTypes.func.isRequired,
  roleInfo: PropTypes.shape().isRequired,
  history: PropTypes.shape().isRequired,
  engineer: PropTypes.shape().isRequired,
  projectId: PropTypes.string.isRequired,
  roleId: PropTypes.string.isRequired,
  projectTitle: PropTypes.string.isRequired,
  cycleId: PropTypes.number.isRequired
};

export default RoleApplication;
