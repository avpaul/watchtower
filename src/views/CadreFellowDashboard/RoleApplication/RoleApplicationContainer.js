import { connect } from 'react-redux';
import RoleApplication from './RoleApplication';
import applyForRole from '../../../redux/actionCreators/applyForRoleActions';

const mapStateToProps = ({ cadreVacancies }) => ({
  applications: cadreVacancies
});

export default connect(
  mapStateToProps,
  {
    applyForRole
  }
)(RoleApplication);
