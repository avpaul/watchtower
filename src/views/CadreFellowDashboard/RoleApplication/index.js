import RoleApplicationContainer from './RoleApplicationContainer';

export default RoleApplicationContainer;
