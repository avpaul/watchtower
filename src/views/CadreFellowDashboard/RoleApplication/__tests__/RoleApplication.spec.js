import React from 'react';
import { shallow, mount } from 'enzyme';
import RoleApplication from '../RoleApplication';
import Loader from '../../../../components/Loader/Loader';
import Fixtures from '../../../../__mocks__/roleApplicationData.json';

jest.useFakeTimers();

describe('tests RoleApplication', () => {
  const defaultProps = {
    roleId: '1',
    loading: false,
    error: null,
    newApplication: null,
    applyForRole: jest.fn(),
    roleInfo: {
      name: 'Technical Team Lead'
    },
    history: {},
    engineer: {},
    projectId: '1',
    projectTitle: 'Project Watchtower'
  };

  const setup = (propOverides = {}, shouldMount = false) => {
    const props = { ...defaultProps, ...propOverides };
    const wrapper = shouldMount
      ? mount(
          <div>
            <button
              type="button"
              data-toggle="modal"
              data-target="applyForRoleModal"
            >
              Send Application
            </button>
            <RoleApplication {...props} />
          </div>
        )
      : shallow(<RoleApplication {...props} />);
    return { props, wrapper };
  };

  beforeEach(() => {
    defaultProps.applyForRole = jest.fn();
    jest.runAllTimers();
  });

  it('should Render Components', () => {
    const wrapper = setup();
    expect(wrapper).toMatchSnapshot();
  });

  it('should not submit form if user inputs whitespaces into the text box', () => {
    const modal = mount(<RoleApplication {...defaultProps} />);
    const instance = modal.instance();
    modal.state().inputs.description.handleValueChange({
      target: { value: 'Hello    ' }
    });

    instance.handleSubmit();
    expect(defaultProps.applyForRole).toHaveBeenCalledTimes(0);
  });

  it('should submit form if user input is okay', () => {
    const modal = mount(<RoleApplication {...defaultProps} />);
    const instance = modal.instance();
    modal.state().inputs.description.handleValueChange({
      target: { value: Fixtures.validApplication_reason }
    });

    instance.handleSubmit();
    expect(defaultProps.applyForRole).toHaveBeenCalledTimes(1);
  });

  it('should test modal close by calling handleClose', () => {
    const { wrapper } = setup(
      {
        applyForRole: jest.fn(),
        roleId: 1,
        projectId: 1,
        history: { replace: jest.fn() }
      },
      true
    );
    const modal = wrapper.find(RoleApplication);
    const buttons = wrapper.find('#applicationBtn');
    buttons.at(0).simulate('click');
    modal.instance().handleClose();
    expect(modal.state('success')).toEqual(false);
  });

  it('should update state on success', () => {
    const modal = mount(<RoleApplication {...defaultProps} />);
    const applications = {
      error: null,
      data: {
        message: 'Success'
      }
    };

    modal.setProps({ applications });
    const instance = modal.instance();
    jest.spyOn(instance, 'applicationStatus');
    modal.instance().applicationStatus(applications.error);
    expect(modal.state('success')).toBe(true);
  });

  it('should update state on failure', () => {
    const modal = mount(<RoleApplication {...defaultProps} />);
    const applications = {
      error: {
        application_reason: [
          'The reason for applying should not be more than 500 characters'
        ]
      },
      data: {}
    };

    modal.setProps({ applications });
    const instance = modal.instance();
    jest.spyOn(instance, 'applicationStatus');
    modal.instance().applicationStatus(applications);
    expect(modal.state('success')).toBe(false);
  });

  it('should update state on failure on attempt to apply with a role', () => {
    const modal = mount(<RoleApplication {...defaultProps} />);
    const applications = {
      error: 'You have an active role',
      data: {}
    };

    modal.setProps({ applications });
    const instance = modal.instance();
    jest.spyOn(instance, 'applicationStatus');
    modal.instance().applicationStatus(applications.error);
    expect(modal.state('errorMessage')).toBe('You have an active role');
  });

  it('should display loader when loading', () => {
    defaultProps.loading = true;
    const modal = mount(<RoleApplication {...defaultProps} />);
    const loader = modal.find(Loader);
    expect(loader.length).toEqual(1);
  });

  it('should test componentDidUpdate', () => {
    const modal = mount(<RoleApplication {...defaultProps} />);
    const instance = modal.instance();
    jest.spyOn(instance, 'applicationStatus');
    modal.setProps({
      applications: { loading: false, data: { message: 'Success' } },
      newApplication: '1-1'
    });
    expect(instance.applicationStatus).toHaveBeenCalled();
  });

  it('should render the relevant feedback message when the user input is less than the acceptable length', () => {
    const modal = mount(<RoleApplication {...defaultProps} />);
    modal.state().inputs.description.handleValueChange({
      target: { value: Fixtures.validApplication_reason }
    });
    expect(modal.find('.feedback-text').text()).toEqual(
      `C'mon, don't sell yourself short!. 😏`
    );
  });

  it('should render the relevant feedback message when the user input is more than the acceptable length', () => {
    const modal = mount(<RoleApplication {...defaultProps} />);
    modal.find('textarea').simulate('change', {
      target: { value: Fixtures.invalidApplication_reason }
    });
    const currentValue = modal.state().inputs.description.getValue();
    modal.instance().evaluateLength(currentValue);

    expect(modal.find('.feedback-text').text()).toEqual(
      `OK, maybe not your full CV, 10x Engineer. 😜`
    );
  });
});
