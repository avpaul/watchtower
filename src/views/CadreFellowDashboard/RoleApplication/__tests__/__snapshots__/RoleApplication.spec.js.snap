// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests RoleApplication should Render Components 1`] = `
Object {
  "props": Object {
    "applications": Object {
      "data": Object {},
      "error": null,
      "loading": false,
    },
    "applyForRole": [MockFunction],
    "roleId": 1,
    "roleInfo": Object {
      "name": "Technical Team Lead",
    },
  },
  "wrapper": <GenericModal
    footer={
      <div
        className="modal-footer"
      >
        <button
          className="btn"
          data-dismiss="modal"
          type="button"
        >
          Cancel
        </button>
        <button
          className="btn"
          id="applicationBtn"
          onClick={[Function]}
          type="button"
        >
          Send Application
        </button>
      </div>
    }
    handleClose={[Function]}
    handleSubmit={[Function]}
    id="applyForRoleModal"
    size=""
    submitLoading={false}
    success={false}
    successMessage="Success!"
    title="Apply to WatchTower Extended Team"
  >
    <div
      className="advocacy"
    >
      <p
        className="advocacy__label"
      >
        You are applying as a Technical Team Lead
      </p>
      <p
        className="advocacy__subtitle"
      >
        Sell yourself (Optional)
      </p>
      <TextInput
        alertText=""
        className="wt-text-input"
        comment=""
        componentStateKey="inputs"
        defaultStatus={0}
        inputValue=""
        label=""
        length={999}
        name="description"
        parent={
          RoleApplication {
            "applicationStatus": [Function],
            "context": Object {},
            "footer": [Function],
            "handleClose": [Function],
            "handleSubmit": [Function],
            "props": Object {
              "applications": Object {
                "data": Object {},
                "error": null,
                "loading": false,
              },
              "applyForRole": [MockFunction],
              "roleId": 1,
              "roleInfo": Object {
                "name": "Technical Team Lead",
              },
            },
            "refs": Object {},
            "renderBody": [Function],
            "renderButton": [Function],
            "sendApplicationButton": [Function],
            "setState": [Function],
            "state": Object {
              "inputs": Object {},
              "success": false,
            },
            "updater": Updater {
              "_callbacks": Array [],
              "_renderer": ReactShallowRenderer {
                "_context": Object {},
                "_didScheduleRenderPhaseUpdate": false,
                "_dispatcher": Object {
                  "readContext": [Function],
                  "useCallback": [Function],
                  "useContext": [Function],
                  "useDebugValue": [Function],
                  "useEffect": [Function],
                  "useImperativeHandle": [Function],
                  "useLayoutEffect": [Function],
                  "useMemo": [Function],
                  "useReducer": [Function],
                  "useRef": [Function],
                  "useState": [Function],
                },
                "_element": <RoleApplication
                  applications={
                    Object {
                      "data": Object {},
                      "error": null,
                      "loading": false,
                    }
                  }
                  applyForRole={[MockFunction]}
                  roleId={1}
                  roleInfo={
                    Object {
                      "name": "Technical Team Lead",
                    }
                  }
                />,
                "_firstWorkInProgressHook": null,
                "_forcedUpdate": false,
                "_instance": [Circular],
                "_isReRender": false,
                "_newState": null,
                "_numberOfReRenders": 0,
                "_renderPhaseUpdates": null,
                "_rendered": <GenericModal
                  footer={
                    <div
                      className="modal-footer"
                    >
                      <button
                        className="btn"
                        data-dismiss="modal"
                        type="button"
                      >
                        Cancel
                      </button>
                      <button
                        className="btn"
                        id="applicationBtn"
                        onClick={[Function]}
                        type="button"
                      >
                        Send Application
                      </button>
                    </div>
                  }
                  handleClose={[Function]}
                  handleSubmit={[Function]}
                  id="applyForRoleModal"
                  size=""
                  submitLoading={false}
                  success={false}
                  successMessage="Success!"
                  title="Apply to WatchTower Extended Team"
                >
                  <React.Fragment>
                    <div
                      className="advocacy"
                    >
                      <p
                        className="advocacy__label"
                      >
                        You are applying as a Technical Team Lead
                      </p>
                      <p
                        className="advocacy__subtitle"
                      >
                        Sell yourself (Optional)
                      </p>
                      <TextInput
                        alertText=""
                        className="wt-text-input"
                        comment=""
                        componentStateKey="inputs"
                        defaultStatus={0}
                        inputValue=""
                        label=""
                        length={999}
                        name="description"
                        parent={[Circular]}
                        placeholder=""
                        testInput={[Function]}
                        type="textarea"
                      />
                    </div>
                  </React.Fragment>
                </GenericModal>,
                "_rendering": false,
                "_updater": [Circular],
                "_workInProgressHook": null,
              },
            },
          }
        }
        placeholder=""
        testInput={[Function]}
        type="textarea"
      />
    </div>
  </GenericModal>,
}
`;
